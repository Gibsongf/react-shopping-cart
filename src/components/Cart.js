import { useEffect, useState } from "react";

export const Cart = () => {
    const [storage, setStorage] = useState({});
    const [quantity, setQuantity] = useState(0);

    const individualProductTotal = (product) => {
        product.total = Number(product.quantity) * Number(product.price);
    };
    const changeProductQuantity = (e) => {
        const product = storage[e.target.id];
        product.quantity = Number(e.target.value);
        setStorage(() => {
            individualProductTotal(product);
            return { ...storage, [product.name]: product };
        });
    };
    const updateStorage = (product) => {
        setStorage(() => {
            individualProductTotal(product);
            return { ...storage, [product.name]: product };
        });
    };

    useEffect(() => {
        const sumQuantity = () => {
            const keys = Object.keys(storage);

            const quantity = keys.map((k) => storage[k].quantity);
            const totalQuantity = quantity.reduce(
                (total, currentNum) => total + currentNum,
                0
            );
            // const parsedNumber = parseInt(totalQuantity.toString(), 10);
            return totalQuantity;
        };

        setQuantity(sumQuantity());
    }, [storage]);

    return { updateStorage, storage, quantity, changeProductQuantity };
};
