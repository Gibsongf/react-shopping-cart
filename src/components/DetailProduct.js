import { useState } from "react";
import "../styles/DetailProduct.css";
import CurrencyIcon from "../imgs/Gold_Currency_Icon.png";
import BackIcon from "../imgs/go-back.png";
import { NavLink } from "react-router-dom";

export const ProductDetails = (props) => {
    const { name, description, imgSrc, price, addToCart } = props;
    const [val, setVal] = useState(1);
    const changeQuantity = (e) => {
        setVal(e.target.value);
    };
    const sendToCart = () => {
        const product = {
            name,
            description,
            price,
            quantity: Number(val),
            imageSrc: imgSrc,
        };
        addToCart(product);
    };
    const GoBack = (
        <NavLink to="/shop">
            <img className="back" src={BackIcon} alt="back-icon" />
        </NavLink>
    );
    return (
        <>
            {props.test ? "" : GoBack}
            <div className="content">
                <img
                    src={imgSrc}
                    alt="item-img-select"
                    className="item-img-select"
                />
                <h2>{name}</h2>
                <p className="description">{description}</p>
                <span className="price-container">
                    <img
                        className="gold-icon-medium"
                        src={CurrencyIcon}
                        alt="gold-coin-medium"
                    />
                    <p>{price}</p>
                </span>
                <div className="add-quantity-cart">
                    <label htmlFor="quantity">
                        Quantity:{" "}
                        <input
                            id="quantity"
                            type="number"
                            min="1"
                            max="100"
                            step="1"
                            value={val}
                            onChange={changeQuantity}
                        />
                    </label>
                    <button onClick={sendToCart} className="addToCart">
                        ADD TO CART
                    </button>
                </div>
            </div>
        </>
    );
};
