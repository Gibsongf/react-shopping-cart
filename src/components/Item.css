
/* a custom font is need here */

.item {
    padding: 10px;
    width: 250px;
    border: solid black 2px;
}
.item-img{
    height: 250px;
    image-rendering: pixelated;
}
.item-img-select{
    height: 300px;
    image-rendering: pixelated;
    border: yellow solid 5px;
}
.close-window{
    width: 24px;
}
#pop-up-content  p{
    font-size: 1.25rem;
    font-weight: 520;

}
#pop-up {
    position: fixed;
    z-index: 1;
    padding-top: 50px;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, 0.4);
}
#pop-up-content {
    margin: 5% auto;
    padding: 20px;
    border-radius: 10px;
    width: 40vh;
    /* #685263 could use this one too  */
    background-color: #ecb1ce;
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
    gap: 10px;
}
.close-window{
    align-self: flex-end;
}
#pop-up-content > span {
    display: flex;
    width: 100%;
    justify-content: space-between;    
}
input[type='number']{
    width: 50px;
}
