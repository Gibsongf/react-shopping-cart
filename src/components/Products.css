.container{
    display: grid;
    grid-template-columns: repeat(auto-fit,minmax(400px,2fr));
    justify-items: center;
    gap: 50px;
}