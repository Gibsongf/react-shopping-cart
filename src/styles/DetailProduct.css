/* a custom font is need here */

@font-face {
    font-family: "main-pixelated";
    src: url("/src/fonts/rainyhearts.woff") format("woff"),
        url("/src/fonts/rainyhearts.woff2") format("woff2");
    ;
}

.item-img-select {
    height: 300px;
    image-rendering: pixelated;
    border: yellow solid 5px;
    padding: 15px;
}

.close-window {
    width: 24px;
}

.content p {
    font-size: 1.25em;
    font-weight: bold;

}

.gold-icon {
    width: 25px;
}


.content {
    margin: 2% auto;
    padding: 20px;
    border-radius: 10px;
    width: 30%;
    background-color: #ecb1ce;
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    flex-direction: column;
    align-items: center;
    text-align: center;
    gap: 10px;

}

.addToCart {
    width: 50%;
    font-family: "main-pixelated";
    font-weight: bold;
    font-size: 1em;

}

.back {
    width: 5%;
    height: auto;
}

.add-quantity-cart {
    display: flex;
    width: 100%;
    justify-content: space-between;
    font-weight: bold;

}

input[type='number'] {
    width: 50px;
}


.price-container {
    display: flex;
    width: 100%;
    align-items: center;
    justify-content: center;
    gap: 10px;
}