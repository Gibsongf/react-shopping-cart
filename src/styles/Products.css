.container{
    display: flex;
    justify-items: center;
    flex-wrap: wrap;
    justify-content: center;
    
}
.item {
    margin: 25px;
}