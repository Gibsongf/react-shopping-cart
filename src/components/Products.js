import "../styles/Products.css";
import CurrencyIcon from "../imgs/Gold_Currency_Icon.png";
import { NavLink } from "react-router-dom";
export const Product = (props) => {
    const { name, imgSrc, price } = props;

    const handleClick = () => {
        // props.clickedItem({ ...props });
        localStorage.setItem("selected-product", JSON.stringify(props));
    };

    return (
        <NavLink
            className="item"
            id="item"
            onClick={handleClick}
            to={`${props.id}`}
            name="item"
        >
            <img src={imgSrc} alt={name} className="item-img" />
            <div className="price-name">
                <h4>{name}</h4>
                <img className="gold-icon" src={CurrencyIcon} alt="gold-coin" />
                {price}
            </div>
        </NavLink>
    );
};
