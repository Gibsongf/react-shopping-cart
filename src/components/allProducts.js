import {
	brutalItems,
	survivalItems,
	tacticsItems,
	allImg,
} from "../getAllImgs.js";
import { Product } from "./Items";
import "./Products.css";

export const AllProducts = () => {
	const brutal = brutalItems();
	const tactics = tacticsItems();
	const survival = survivalItems();

	const tacticsImgs = allImg(tactics, "tactics");
	const brutalImgs = allImg(brutal, "brutality");
	const survivalImg = allImg(survival, "survival");

	const arrayAll = [brutalImgs, tacticsImgs, survivalImg];
	const allProducts = arrayAll.map((imgSet) =>
		Object.keys(imgSet).map((fKey) => {
			return (
				<Product
					key={fKey}
					id={imgSet[fKey].id}
					imgSrc={imgSet[fKey].src}
					description={imgSet[fKey].description}
					name={imgSet[fKey].name}
					price={imgSet[fKey].price}
				/>
			);
		})
	);
	return allProducts;
};
