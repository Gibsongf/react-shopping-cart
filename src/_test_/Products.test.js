import { render, screen } from "@testing-library/react";
import React from "react";
import { Product } from "../components/Products";
import { ProductDetails } from "../components/DetailProduct";
import Src from "../imgs/tactics/Alchemic_Carbine_Icon.png";
import userEvent from "@testing-library/user-event";

// const ProductSetup = (props) => {
// 	const name = "The Alchemic Carbine";
// 	const description =
// 		"Crossbow-type ranged weapon which poisons enemies hit by its bolts.";
// 	const price = 2000;

// 	return (
// 		<ProductDetails
// 		name={name}
// 		description={description}
// 		price={price}
// 		imgSrc={Src}
// 		addToCart={props.addToCart}
// 		test={true}
// 		/>
// 	);
// };

// describe("open and close window of the product", () => {
// 	it('renders product details',()=>{
// 		const { container } = render(<ProductSetup />);
//     	expect(container).toMatchSnapshot();
// 	})

// 	it("add to cart called right", async () => {
// 		const addCartMock = jest.fn()
// 		const user = userEvent.setup()
// 		render(<ProductSetup addToCart={addCartMock}/>);
// 		const btnAdd = screen.getByRole("button", { name: "ADD TO CART" });
// 		await user.click(btnAdd)
// 		expect(addCartMock).toHaveBeenCalledTimes(1);
// 	});
// });
