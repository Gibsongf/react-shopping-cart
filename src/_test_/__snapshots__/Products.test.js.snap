// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`open and close window of the product renders product details 1`] = `
<div>
  <div
    class="content"
  >
    <img
      alt="item-img-select"
      class="item-img-select"
      src="Alchemic_Carbine_Icon.png"
    />
    <h2>
      The Alchemic Carbine
    </h2>
    <p
      class="description"
    >
      Crossbow-type ranged weapon which poisons enemies hit by its bolts.
    </p>
    <span
      class="price-container"
    >
      <img
        alt="gold-coin"
        class="gold-icon"
        src="Gold_Currency_Icon.png"
      />
      <p>
        2000
      </p>
    </span>
    <div
      class="add-quantity-cart"
    >
      <label
        for="quantity"
      >
        Quantity:
         
        <input
          id="quantity"
          max="100"
          min="1"
          step="1"
          type="number"
          value="1"
        />
      </label>
      <button
        class="addToCart"
      >
        ADD TO CART
      </button>
    </div>
  </div>
</div>
`;
